### DISCUSSION TITLE
How to structure code

### DISCUSSION THREAD

**Discussion by G3UKB** [emilk/egui#2236] [Q&A]
I've been trying various ways to structure the gui code but pretty much failing. The basic eframe examples are all pretty, well basic. The web demo is impressive but I can't figure the code behind the demo code. I installed the template but that also puts the entire gui in update(). In the template what is the best way to put each of those panels into a method of TemplateApp whilst retaining an &mut self parameter (that's the bit I can't figure). I can move content into a stand alone function (no good) or a method of TemplateApp without a self parameter (no good). Self is required to save state and communicate between widgets. This is done in the web demo code but behind the scenes which is too complex for me to follow at the moment.

Hopefully this is an easy question to answer as my previous question on mouse events hasn't been answered and as I think its probably not possible I was looking for an alternative way to do what I need to do and ran aground on communication between widgets. Hence this question. Thanks for any assistance.

**Comment by n00kii** [comment]
the way i do it is have a unique struct for every distinct page of the ui (like in the demo), and have them `impl` a `UserInterface` trait. that way i can keep all those different structs a `Vec` of trait objects under the main app struct, and call `.ui` (from the trait) on them to conditionally render. 

for state, i recommend having a field in the main app struct keeps an `Rc<RefCell<State>>` where `State` is whatever struct u wanna use for keeping shared state/information, and then `Rc::clone`-ing the state variable to whatever other ui structs need it

**Reply by G3UKB** [reply]
Thanks. That sounds like a good arrangement. It will take me a while to get my head around that but nice to know it can have a good structure. I've tried all the main GUI's now and pretty much discarded the others as missing features I needed or just plain too difficult to structure or too complicated to use. I think immediate mode is probably the best fit for Rust and egui looks really nice too.
